audio/filter/af_scaletempo.c

```diff
@@ -135,72 +134,144 @@ static bool fill_queue(struct priv *s)
+        history[0] = history[1];
+        history[1] = history[2];
+        history[2] = distance;
+        if(offset >= 2 && history[0] >= history[1] && history[1] <= history[2]) {
+            float extremum;
+            quadratic_interpolation_float(history, &extremum, &distance);
+            offset_approx = offset - step_size + (int)(extremum * step_size + 0.5f);
+        }
+
+        if (distance < best_distance) {
+            best_distance = distance;
+            best_offset_approx  = offset_approx;
+        }
+    }
+
+    best_distance = FLT_MAX;
+    int best_offset = 0;
+    int min_offset = MPMAX(0, best_offset_approx - step_size + 1);
+    int max_offset = MPMIN(frames_search, best_offset_approx + step_size);
+    for (int offset = min_offset; offset < max_offset; offset++) {
+        float distance = 0;
+        for (int i = 0; i < num_samples; i++)
+            distance += fabs(target[i] - source[offset * num_channels + i]);
+        if (distance < best_distance) {
+            best_distance = distance;
+            best_offset  = offset;
         }
-        search_start += s->num_channels;
     }
 
-    return best_off * 4 * s->num_channels;
+    return best_offset * 4 * num_channels;
 }
 
 static int best_overlap_offset_s16(struct priv *s)
 {
-    int64_t best_corr = INT64_MIN;
-    int best_off = 0;
-
-    int32_t *pw  = s->table_window;
-    int16_t *po  = s->buf_overlap;
-    po += s->num_channels;
-    int32_t *ppc = s->buf_pre_corr;
-    for (long i = s->num_channels; i < s->samples_overlap; i++)
-        *ppc++ = (*pw++ **po++) >> 15;
-
-    int16_t *search_start = (int16_t *)s->buf_queue + s->num_channels;
-    for (int off = 0; off < s->frames_search; off++) {
-        int64_t corr = 0;
-        int16_t *ps = search_start;
-        ppc = s->buf_pre_corr;
-        ppc += s->samples_overlap - s->num_channels;
-        ps  += s->samples_overlap - s->num_channels;
-        long i  = -(s->samples_overlap - s->num_channels);
-        do {
-            corr += ppc[i + 0] * (int64_t)ps[i + 0];
-            corr += ppc[i + 1] * (int64_t)ps[i + 1];
-            corr += ppc[i + 2] * (int64_t)ps[i + 2];
-            corr += ppc[i + 3] * (int64_t)ps[i + 3];
-            i += 4;
-        } while (i < 0);
-        if (corr > best_corr) {
-            best_corr = corr;
-            best_off  = off;
+    int num_channels = s->num_channels, frames_search = s->frames_search;
+    int16_t *source = (int16_t *)s->buf_queue + num_channels;
+    int16_t *target = (int16_t *)s->buf_overlap + num_channels;
+    int num_samples = s->samples_overlap - num_channels;
+    int step_size = 3;
+    int32_t history[3] = {};
+
+    int32_t best_distance = INT32_MAX;
+    int best_offset_approx = 0;
+    for (int offset = 0; offset < frames_search; offset += step_size) {
+        int32_t distance = 0;
+        for (int i = 0; i < num_samples; i++)
+            distance += abs(target[i] - source[offset * num_channels + i]);
```

Comment:
Should we cast to wider type to avoid underflow?
```suggestion
            distance += abs((int32_t)target[i] - (int32_t)source[offset * num_channels + i]);
```
and similar in other places

Reply:
Yes we should do that.